### PR TITLE
Multi mtp magic send

### PR DESCRIPTION
**Pull request by FeixLiu**
### PR Category

Performance Optimization

### PR Types

New features

### Description

Multi mtp magic send

### 是否引起精度变化

否

### CODE REVIEW

File: src/paddlefleet/transformer/multi_token_prediction.py [hunk 1]
```diff
@@ -774,57 +808,76 @@ def forward(self, dict_args: dict):
...
+            from paddlefleet.models.gpt.mtp_embedding_layer import (
+                mtp_magic_instance,
+            )
+
+            magic_count = mtp_magic_instance.get_magic_count(self.magic_key)
+            # Only increment on the real forward pass, not during recompute replay.
+            if paddle.is_grad_enabled() or not self.training:
+                magic_count += 1
+                mtp_magic_instance.set_magic_count(self.magic_key, magic_count)
+            input_ids_list = mtp_magic_instance.get("input_ids")
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/multi_token_prediction.py:841]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这里切到 `mtp_magic_instance.get("input_ids")` 后，当前仓库没有任何路径会填充这个 key。`MagicInstance.set_data()` 只有定义没有调用；同时本 PR 又把原先插入模型图的 `MTPEmbeddingLayer` 去掉了，旧的 `input_ids_for_mtp` deque 也不再被新的 forward 使用。因此只要开启 `enable_mtp_magic_send`，第一次走到这里就会触发 `input_ids not in magic_send_dict`，MTP magic-send 功能无法运行。

请把输入侧广播/缓存接回新的 `MagicInstance`，或者继续沿用一个实际会被生产者填充的输入通道，并补一个覆盖 `enable_mtp_magic_send=True` 的 forward 测试。修复形态可以类似：

```python
# 在每个 accumulation/eval step 的输入广播完成后执行
broadcasted = broadcast_data_obj({"input_ids": input_ids_list}, src_rank=..., group=...)
mtp_magic_instance.set_data({"input_ids": broadcasted["input_ids"]})
mtp_magic_instance.clear_count_dict()
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
发现一个需要修复后再合入的问题，细节已放在行级评论里。

- P3 **优先级：P3**
  非行级：PR 标题和描述不属于 diff 可评论行。当前标题/描述只有 “Multi mtp magic send”，没有说明这次变更实际涉及多层 MTP magic-send、输入广播缓存、embedding 权重同步和配置约束放宽，也缺少测试/风险说明。**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。** 建议标题改成类似 “Support multi-layer MTP magic send”，并在 Description 里补充主要实现点、覆盖的 PP/TP/VPP 组合测试以及已知限制。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/models/gpt/mtp_embedding_layer.py [hunk 2]
```diff
@@ -12,101 +12,47 @@
...
 from __future__ import annotations
 
-import copy
-from collections import deque
-from typing import TYPE_CHECKING
 
-from paddlefleet.models.gpt.utils import fill_feature
-from paddlefleet.tensor_parallel import VocabParallelEmbedding
-from paddlefleet.transformer.layer import FleetLayer
+class MagicInstance:
```
Comment (risemeup1111) [inline src/paddlefleet/models/gpt/mtp_embedding_layer.py:24]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这个文件现在只保留了 `MagicInstance`，但现有 `tests/single_card_tests/transformer/test_mtp_magic_send.py` 仍在模块导入阶段执行：

```python
from paddlefleet.models.gpt.mtp_embedding_layer import (
    MTPEmbeddingLayer,
    input_ids_for_mtp,
)
```

这两个符号在本 commit 中被删除后，单测收集时会直接 `ImportError`，后面的测试不会开始执行。请同步更新测试（删除/改写 `TestMTPEmbeddingLayer`，并改成覆盖当前 `mtp_magic_instance` 的数据写入与读取路径），或恢复兼容导出直到测试和调用方全部迁移完成。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
复查了新的提交。上轮关于 `mtp_magic_instance` 没有输入生产者的 P1 仍未解决，原行级线程继续作为该问题的讨论位置；这次新增提交还引入了一个测试导入失败的问题，已在新的行级评论里说明。

- P3 **优先级：P3**
  非行级：PR 标题和描述仍然只有 “Multi mtp magic send”，还没有补充主要实现点、测试覆盖和风险/限制说明。**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。** 建议标题改为类似 “Support multi-layer MTP magic send”，并在 Description 中补充本次支持的 PP/TP/VPP 组合、输入广播缓存方案、权重同步方案和测试计划。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/multi_token_prediction.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/multi_token_prediction.py:841]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

复查到最新提交后，这个阻塞点仍未解决：当前新增的 `_setup_magic()` 只在单测里调用 `mtp_magic_instance.set_data({"input_ids": ...})`，全仓生产代码里仍然只有 `MultiTokenPredictionLayer` 在读取 `mtp_magic_instance.get("input_ids")`，没有训练/推理数据流在 magic-send forward 前写入这个 key。也就是说单测通过手工注入绕过了真实输入侧生产路径，实际开启 `enable_mtp_magic_send` 时仍会触发 `input_ids not in magic_send_dict`。

请把输入侧广播/缓存逻辑接到生产代码路径，例如在真实 micro-batch/accumulation 数据广播完成后写入：

```python
broadcasted = broadcast_data_obj({"input_ids": input_ids_list}, src_rank=..., group=...)
mtp_magic_instance.set_data({"input_ids": broadcasted["input_ids"]})
mtp_magic_instance.clear_count_dict()
```

同时保留一个不依赖 `_setup_magic()` 手工注入的覆盖用例，验证生产入口会填充 `mtp_magic_instance` 后再进入 MTP forward。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交。测试里删除 `MTPEmbeddingLayer` / `input_ids_for_mtp` 的旧导入后，Nyanpasu 之前关于测试收集 ImportError 的 P1 已由新 commit 修复。

但 `mtp_magic_instance` 的 `input_ids` 生产路径仍未接入生产代码，我已在原行级线程补充了最新证据。这个 P1 仍需要修复并提交新的 commit 后才能继续推进。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/multi_token_prediction.py [hunk 3]
```diff
@@ -838,94 +897,112 @@ def forward(self, dict_args: dict):
...
+                        :, depth : depth + 1, :, :1
+                    ]
+            mtp_hidden_inputs_mask = (
+                mtp_hidden_inputs_mask_all[:, depth : depth + 1, :]
+                if mtp_hidden_inputs_mask_all is not None
+                else None
+            )
+
+            # Projection + transformer
+            _proj_kwargs = {
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/multi_token_prediction.py]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这轮改动把 magic-send 分支改成原地保留 `dict_args`，注释里也写了要保留 `blocks`、`position_ids` 等字段；但这些字段只是在当前 MTP transformer 执行之后才被保留下来。当前这里构造 `_proj_kwargs` 时仍只传 rotary/MTP mask/input_ids，`attention_mask`、`position_ids`、`attention_bias`、`blocks` 等字段没有参与本层 `_proj_and_transformer_layer()` 调用；而 `_proj_and_transformer_layer()` 内部构造给 `self.transformer_layer(input_dict)` 的 `input_dict` 也不会转发 `kwargs` 里的 `blocks`。因此开启 block attention residuals、显式 position_ids/attention_bias 等配置时，当前 MTP 层仍会按空 `blocks` 或 `None` 参数执行，保留下游字段不能修复本层行为。

请把当前 MTP transformer 调用需要的原始参数一起传入，并在 `_proj_and_transformer_layer()` 内转发给 `self.transformer_layer`。修复形态可以类似下面这样，具体切片规则请和非 magic-send MTP 路径保持一致：

```python
_proj_kwargs = {
    "attention_mask": dict_args.get("attention_mask", None),
    "attention_bias": dict_args.get("attention_bias", None),
    "position_ids": trim_or_select_mtp_position_ids(dict_args, depth, seq_len),
    "blocks": dict_args.get("blocks", None),
    "rotary_pos_emb": _trim_rotary("rotary_pos_emb"),
    "rotary_pos_cos": _trim_rotary("rotary_pos_cos"),
    "rotary_pos_sin": _trim_rotary("rotary_pos_sin"),
    "swa_rotary_pos_emb": _trim_rotary("swa_rotary_pos_emb"),
    "swa_rotary_pos_cos": _trim_rotary("swa_rotary_pos_cos"),
    "swa_rotary_pos_sin": _trim_rotary("swa_rotary_pos_sin"),
    "attn_mask_startend_row_indices": mtp_mask,
    "mtp_hidden_inputs_mask": mtp_hidden_inputs_mask,
    "input_ids": mtp_input_ids_local,
}

# inside _proj_and_transformer_layer()
input_dict = {
    ...,
    "blocks": kwargs.get("blocks", None),
}
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交。新的 `dict_args` 解析修复仍不完整，我在行级评论里说明了当前 MTP transformer 调用仍未拿到本层需要的原始参数。

另外，Nyanpasu 之前指出的 `mtp_magic_instance` 缺少生产路径写入 `input_ids` 的 P1 仍未由本次提交修复，原行级线程继续作为该问题的讨论位置。请两个 P1 都修复并提交新的 commit 后再继续推进。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/multi_token_prediction.py (hunk 3 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/multi_token_prediction.py]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

复查最新提交后，这个点只完成了部分修复：`dict_args` 现在会传给 `_proj_and_transformer_layer(**dict_args)`，所以 `attention_mask`、`attention_bias` 等签名参数已经可以进入本层调用。但当前实现仍有两个缺口：

1. `blocks` 仍没有传入内部 `self.transformer_layer(input_dict)`。`_proj_and_transformer_layer()` 虽然接收 `**kwargs`，但构造 `input_dict` 时没有加入 `"blocks": kwargs.get("blocks", None)`，开启 `block_attention_residuals` 时内部 `TransformerLayer.forward()` 仍会走空 `blocks`，MTP 层没有继承前面累积的 block 状态。
2. `position_ids` 现在是原样透传。magic-send 分支的 `decoder_input` 已经按 `(depth + 1):(depth + 1 + seq_len)` 取当前 MTP depth 的窗口，显式 `position_ids` / 推理 MRoPE 场景也需要按同一窗口选择当前 depth 的位置，而不是把原始字段直接交给当前 MTP transformer。

建议把这两个字段补齐到当前层真实调用路径，并加一个 magic-send 覆盖用例验证 `_proj_and_transformer_layer()` 收到的 `blocks` 和当前 depth 的 `position_ids`：

```python
# before _proj_and_transformer_layer
if "position_ids" in dict_args and dict_args["position_ids"] is not None:
    dict_args["position_ids"] = select_mtp_position_ids(
        dict_args["position_ids"], depth, seq_len
    )

# inside _proj_and_transformer_layer()
input_dict = {
    ...,
    "position_ids": position_ids,
    "blocks": kwargs.get("blocks", None),
}
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交。`dict_args` 透传修复有进展，但 Nyanpasu 之前指出的 MTP transformer 参数透传 P1 仍未完全解决，我已在原行级线程补充当前剩余缺口。

另外，`mtp_magic_instance` 的 `input_ids` 生产路径仍未接入生产代码，原行级线程继续作为该 P1 的讨论位置。请这两个 P1 都修复并提交新的 commit 后再继续推进。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/multi_token_prediction.py [hunk 4]
```diff
@@ -838,96 +895,183 @@ def forward(self, dict_args: dict):
...
+            if self.config.sequence_parallel:
+                # [B, S] -> [B, S/tp]: flatten, scatter, reshape back
+                B_local = mtp_input_ids_local.shape[0]
+                mtp_input_ids_local = mtp_input_ids_local.reshape(
+                    [-1]
+                ).unsqueeze(-1)
+                mtp_input_ids_local = ScatterOp.apply(mtp_input_ids_local)
+                mtp_input_ids_local = mtp_input_ids_local.squeeze(-1).reshape(
+                    [B_local, -1]
+                )
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/multi_token_prediction.py:920]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这里把 `mtp_input_ids_local` 先按 CP scatter 后再传给内部 Transformer，会和 MoE router 的现有逻辑重复切分。`TopKRouter.forward()` 在 experimental_dataflow 下本来就会自己做 CP scatter；而 `gpt_model_use_experimental_version=True` 的 2D 分支还会额外做 SP scatter。现在这段代码会让 `input_ids` 先少一截，再被 router 再切一次，最终和 `hidden_states` 的局部形状对不上，MTP MoE 路径会直接断言失败。

请把这里的预处理收回到 router 没有覆盖的那一部分：CP 相关的切分不要在这里做，只保留非 experimental 3D 路径需要的 SP 预切分。可直接改成：

```suggestion
            if (
                self.config.sequence_parallel
                and not self.config.gpt_model_use_experimental_version
            ):
                B_local = mtp_input_ids_local.shape[0]
                mtp_input_ids_local = mtp_input_ids_local.reshape(
                    [-1]
                ).unsqueeze(-1)
                mtp_input_ids_local = ScatterOp.apply(mtp_input_ids_local)
                mtp_input_ids_local = mtp_input_ids_local.squeeze(-1).reshape(
                    [B_local, -1]
                )
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交，仍有阻塞问题，细节已在 inline 评论里。上一轮提出的输入侧写入问题目前仍未在生产路径上解决。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by From00 — APPROVED** [review]
LGTM
